cloudflare/resource_cloudflare_access_application_test.go

```diff
@@ -95,6 +95,33 @@ func TestAccCloudflareAccessApplicationWithAutoRedirectToIdentity(t *testing.T)
 	})
 }
 
+func TestAccCloudflareAccessApplicationWithADefinedIdps(t *testing.T) {
+	accountID := os.Getenv("CLOUDFLARE_ACCOUNT_ID")
+	rnd := generateRandomResourceName()
+	name := fmt.Sprintf("cloudflare_access_application.%s", rnd)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+		},
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckCloudflareAccessApplicationDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCloudflareAccessApplicationConfigWithADefinedIdp(rnd, zoneID, domain, accountID),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(name, "zone_id", zoneID),
+					resource.TestCheckResourceAttr(name, "name", rnd),
+					resource.TestCheckResourceAttr(name, "domain", fmt.Sprintf("%s.%s", rnd, domain)),
+					resource.TestCheckResourceAttr(name, "session_duration", "24h"),
+					resource.TestCheckResourceAttr(name, "auto_redirect_to_identity", "true"),
+					resource.TestCheckResourceAttr(name, "allowed_idps.#", "2"),
```

Comment:
@jacobbednarz maybe you can help, me with that one. I don't know how I could check if the value inside `allowed_idps.0` is the same as the one created at line 169 (cloudflare_access_identity_provider.id) ?

Reply:
As `schema.TypeList` is an [ordered list](https://www.terraform.io/docs/extend/schemas/schema-types.html#typelist) you should be able to check `allowed_idps.0` for your initial value if it's being set correctly. `allowed_idps.# = 2` seems a little odd here as you've only got a single entry in your test below.

Reply:
Good catch, `allowed_idps.# = 2` was indeed a typo, I corrected it.

Regarding [the ordered list](https://www.terraform.io/docs/extend/schemas/schema-types.html#typelist), `allowed_idps.0` in my case should be equal to the UUID (that I don't know prior the resource creation) of https://github.com/terraform-providers/terraform-provider-cloudflare/pull/734/commits/dfb005085a8c1e06239c5eece6ff4d669303f3e3#diff-d5035ca3592bc21fc69fafd3d40e2145R169.

```
resource "cloudflare_access_identity_provider" "%[1]s" {
  account_id = "%[4]s"
  name = "%[1]s"
  type = "onetimepin"
```

So I'm wondering how I could test this properly ?

Reply:
Instead of using `TestCheckResourceAttr`, you can use [`TestCheckResourceAttrSet`](https://godoc.org/github.com/hashicorp/terraform-plugin-sdk/helper/resource#TestCheckResourceAttrSet) which doesn't evaluate the key value but instead just the presence of it. An example would be:

```
resource.TestCheckResourceAttrSet(name, "allowed_idps.0"),
```

This would check that `allowed_idps.0` is set irrespective of the value it holds.

Another option (or could be additional test!) is to check the value with a regex using [`TestMatchResourceAttr`](https://godoc.org/github.com/hashicorp/terraform-plugin-sdk/helper/resource#TestMatchResourceAttr). Example:

```
resource.TestMatchResourceAttr(name, "allowed_idps.0", regexp.MustCompile("^[a-fA-F0-9]{32}$"))
```

Reply:
All of these test methods are defined at https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/resource/testing.go so feel free to peruse and find the one that makes the most sense here.

Reply:
Perfect thanks a lot for the pointers !